fix(ngcc): several ngcc fixes and test improvements

gkalpak opened the pull request:
See individual commits for details.

##
Fixes #44019.

mary-poppins commented:
You can preview 8e4d73c at https://pr44245-8e4d73c.ngbuilds.io/.

mary-poppins commented:
You can preview b4e0495 at https://pr44245-b4e0495.ngbuilds.io/.

mary-poppins commented:
You can preview b6a1732 at https://pr44245-b6a1732.ngbuilds.io/.

mary-poppins commented:
You can preview bae78d7 at https://pr44245-bae78d7.ngbuilds.io/.

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -547,51 +551,216 @@ function getUmdWrapper(statement: ts.Statement):
+        stmt.getText());
+  }
+
+  return factoryCalls as (typeof factoryCalls&{commonJs: ts.CallExpression});
```
Would it make sense to do:

```suggestion
  const commonJs = getCommonJsFactoryCall(conditionalFactoryCalls);
   if (commonJs === null) {
     throw new Error(
         'Unable to find a CommonJS factory call inside the UMD wrapper function:\n' +
         stmt.getText());
   }

   return {
     amdDefine: getAmdDefineCall(conditionalFactoryCalls),
     commonJs,
     global: getGlobalFactoryCall(conditionalFactoryCalls),
   };
```

that avoids the slightly awkward cast in the return position.

**EDIT** Ah, the subsequent commit may explain why it's done like this; perhaps the above structure doesn't apply that nicely in that case anymore.

gkalpak replied:
Yeah, this might work for one property, but not for two (as we need in a subsequent commit).
I couldn't find a way to implement this in a way that TS understands that "at least one of two properties is non-null" (without type-casting) 😞

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -547,51 +551,216 @@ function getUmdWrapper(statement: ts.Statement):
+function getRequiredModulePath(cjsFactoryCall: ts.CallExpression, paramIndex: number): string {
+  const requireCall = cjsFactoryCall.arguments[paramIndex];
+
+  if (!isRequireCall(requireCall)) {
```
WDYT about safe-guarding against the argument for index `paramIndex` missing:

```suggestion
  if (requireCall !== undefined && !isRequireCall(requireCall)) {
```

gkalpak replied:
Good idea :+1:

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -553,10 +553,15 @@ function getUmdWrapper(statement: ts.Statement):
- * The supported format for the UMD wrapper function body is a single statement which is a
- * `ts.ConditionalExpression` (i.e. using a ternary operator). For example:
+ * NOTE:
+ * For more info on the distinction between CommonJS and CommonJS2 see
+ * https://github.com/webpack/webpack/issues/1114.
```
TIL

gkalpak replied:
Yeah, I learned about that while working on this PR too 😅

JoostK commented on packages/compiler-cli/ngcc/test/dependencies/umd_dependency_host_spec.ts:
```diff
@@ -143,6 +152,15 @@ runInEachFileSystem(() => {
+              '// into processing it even if it has no actual dependencies.\n' +
+              umd('no_imports_but_cannot_skip', []),
+        },
+        {name: _('/no/imports/but/cannot/skip/package.json'), contents: '{"esm2015": "./index.js"}'},
```
Noticing it now but this is everywhere in this file: `esm2015` should probably not be used for the UMD bundle.

gkalpak replied:
Yeah, I noticed that too and wasn't sure why it was that way. Possibly a copy/paste carry-over from another `ReflectionHost` test suite. I can try fixing it (and see if anything breaks).

JoostK commented on packages/compiler-cli/ngcc/test/helpers/umd_utils.ts:
```diff
@@ -0,0 +1,124 @@
+        throw new Error(`Unknown UMD wrapper function format: ${wrapperFunctionFormat}`);
+    }
+
+    wrapperFunctionCall = wrapperFunctionCall.replace('<BODY_PLACEHOLDER>', factoryBody);
```
Is this done separately to prevent the body from being subject to `stripIndentation`? Is the indentation of the body significant? Could you add a comment here why this is done?

gkalpak replied:
Yeah, it is to avoid affecting the indentation of the body. It is not significant from a functionality perspective, but changing the indentation would affect some test assertions (which use a specific indentation), so I wanted to avoid unnecessary changes.

I'll add a comment.

JoostK commented on packages/compiler-cli/ngcc/test/helpers/umd_utils.ts:
```diff
@@ -0,0 +1,124 @@
+  const lines = text.replace(/^ *\r?\n/, '').replace(/\r?\n *$/, '').split('\n');
+  const minIndentation =
+      Math.min(...lines.filter(l => !/^ *\r?$/.test(l)).map(l => /^ */.exec(l)![0].length));
+  const re = new RegExp(`^ {0,${minIndentation}}`);
+
+  return lines.map(l => l.replace(re, '')).join('\n');
```
Isn't this equivalent to a simple substring on each line from `minIndentation`? I know you like regexes, but I'm not seeing why this needs them.

JoostK replied:
Wouldn't using `'  '.substring(minIndentation)` avoid that problem? The start index would then exceed the line length, resulting in the empty string.

gkalpak replied:
Ehm...OK...I didn't think about that 😅
That would work 🙈

JoostK commented on packages/compiler-cli/ngcc/test/rendering/umd_rendering_formatter_spec.ts:
```diff
@@ -20,10 +20,14 @@ import {NgccReferencesRegistry} from '../../src/analysis/ngcc_references_registr
-  contents: {preamble?: string; wrapperFunction: string; wrapperCallArguments: string;};
+  contents: {
+    preamble?: string; moduleName: string; dependencies: string[]; factoryBody: string;
+    additionalOptions?: AdditionalFormatOptions,
```
```suggestion
    additionalOptions?: AdditionalFormatOptions;
```

(hopefully that helps clang-format in rendering newlines between all properties)

gkalpak replied:
😱

gkalpak replied:
For the record, no it doesn't help clang-format 🤦‍♂️

JoostK commented on packages/compiler-cli/ngcc/test/rendering/umd_rendering_formatter_spec.ts:
```diff
@@ -235,10 +221,34 @@ exports.BadIife = BadIife;
+                            `    module.exports = factory(require('@angular/core'),require('@angular/common'),require('some-side-effect'), require('/local-dep'), require('@angular/core'));`);
           });
 
+          if (wrapperFunctionFormat === WrapperFunctionFormat.Webpack) {
+            // The CommonJS (vs CommonJS2) format only applies to the Webpack format.
+            it('should append the given imports into the CommonJS2 factory call', () => {
```
I am confused here. The condition enables this test only for Webpack wrappers and the test description tells me that that uses CommonJS2, but I read the comment as the exact opposite (where Webpack only uses CJS)

petebacondarwin replied:
I think this should be 

```suggestion
          if (wrapperFunctionFormat === WrapperFunctionFormat.Webpack) {
            // The CommonJS (vs CommonJS2) format only applies to the Webpack format.
            it('should append the given imports into the CommonJS factory call', () => {
```

We are checking here that in webpack there is also CommonJS (v1) as well as CommonJS2.
The test above this already tests the CommonJS2 format for both rollup and webpack.
Right?

gkalpak replied:
Ooops, yes, it should say `CommonJS` in the test description 😇

JoostK reviewed (approved):
Thanks for the extensive test updates! 💯

gkalpak commented:
FYI, I created angular/ngcc-validation#4955 to test this and it seems that it [can fail][1] when trying to sniff a UMD module with an unsupported wrapper function format (such as [ng-zorro-antd/bundles/ng-zorro-antd-core-types.umd.js][2]).

I'll need to fix this 😞 

[1]: https://circleci.com/gh/angular/ngcc-validation/72149
[2]: https://unpkg.com/browse/ng-zorro-antd@12.0.2/bundles/ng-zorro-antd-core-types.umd.js

JoostK commented:
Interestingly, the error in ngcc-validation highlights that `sniffModuleFormat` also needs to parse the source file with parent pointers enabled.

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/umd_host_spec.ts:
```diff
@@ -1769,7 +1771,7 @@ runInEachFileSystem(() => {
             case 'imported':
               fileHeaderWithUmd = `
               (function (global, factory) {
-                typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('tslib'))) :
```
Were these extra closing parentheses just a mistake in the original tests or needed due to the changes in this commit?

gkalpak replied:
No, they were just mistakes. They didn't matter before, because the tests that use these files did not actually parse the wrapper function (only the factory function).

With the changes in this PR, the wrapper function is processed eagerly (even if not needed), thus the tests were failing.

(Actually, I am going to make the processing of the wrapper function lazy (to deal with [other issues][1]), but I think it is still a good idea to keep the test code here valid, so I am going to keep these changes even if not strictly necessary :grin:)

[1]: https://github.com/angular/angular/pull/44245#issuecomment-975912659

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -719,13 +822,19 @@ function isTypeOf(node: ts.Expression, ...types: string[]): boolean {
     {parameter: ts.ParameterDeclaration, path: string}[] {
   const imports: {parameter: ts.ParameterDeclaration, path: string}[] = [];
-  const cjsFactoryCall = umdModule.factoryCalls.commonJs;
+  const cjsFactoryCall = umdModule.factoryCalls.commonJs2 ?? umdModule.factoryCalls.commonJs!;
```
Given that we are overriding the type system here anyway (by asserting that one of these must be non-null) I don't see any point in complicating the `factoryCalls` type with the `RequireAtLeastOneNonNullable<...>` type transform.

Why not just let the type of `UmdModule` be:

```
interface UmdModule {
  wrapperFn: ts.FunctionExpression;
  factoryFn: ts.FunctionExpression;
  factoryCalls: Record<'amdDefine'|'commonJs'|'commonJs2'|'global', ts.CallExpression|null>;
}
```

and still have the runtime check that at least one of commonjs and commonjs2 are non-null.

gkalpak replied:
My hope was that (at some point in the future) TS would become smart enough to be able to infer that if one is null/undefined then the other is not. Also keeping the type would make it easier for people to understand why the non-null assertion is safe here.

Or are you suggesting moving the runtime check here?

(In any case, I don't feel too strongly about it - esp. given that TS cannot understand it so well - so I would be happy to remove the complicated type if you think it is not worth it.)

petebacondarwin replied:
If it was me, I would not add this type. But I won't block the PR over it.

One thing to consider is that

```ts
umdModule.factoryCalls.commonJs2 ?? umdModule.factoryCalls.commonJs!
```

is really a private implementation concern of the `UmdModule` and not the `UmdReflectionHost` IMO.

So I would consider doing one of two things:

* Keep the type but create a helper function that returns a non-null factory call that cn be used here. This helper would be close to the place where we parse the factoryCalls out of the UMD wrapper.
* Drop the fancy type and have something more like:
  ```ts
  interface UmdModule {
    wrapperFn: ts.FunctionExpression;
    factoryFn: ts.FunctionExpression;
    factoryCalls: Record<'amdDefine'|'commonJs'|'commonJs2'|'global', ts.CallExpression|null>;
    factoryCallForImports: ts.CallExpression;
  }
  ```

gkalpak replied:
OK, I went for the second option. I modified it a bit, making the ``forImports` calls a member of the `factoryCalls` object.
See https://github.com/angular/angular/pull/44245/commits/6448a403d28d91c6551d63fce04a188ce71d7423.

petebacondarwin replied:
Great!!

petebacondarwin commented on packages/compiler-cli/ngcc/test/helpers/umd_utils.ts:
```diff
@@ -0,0 +1,124 @@
+  Webpack = 'wrapper function as emitted by Webpack',
+}
+
+export function createUmdModuleFactory(
```
Call me old fashioned but I feel like this leads better to an OO abstraction, with two subclasses (one for rollup and one for webpack). That would avoid the switch statement.

gkalpak replied:
Oh, you are old fashioned!

petebacondarwin commented on packages/compiler-cli/ngcc/test/helpers/umd_utils.ts:
```diff
@@ -114,6 +114,25 @@ export function createUmdModuleFactory(
   };
 }
 
+export function testForEachUmdFormat(testSuite: (data: {
```
Nice!

petebacondarwin reviewed (approved):
Nothing blocking. Thanks for going above and beyond, as always!

gkalpak commented:
> Interestingly, the error in ngcc-validation highlights that `sniffModuleFormat` also needs to parse the source file with parent pointers enabled.

The error happens now (and didn't happen before), because we eagerly try to parse the UMD module wrapper function. I think in some cases, this will never be needed (for example, we might never need to process an entry-point or it might be a types-only entry-point (as seems to be the case for [ng-zorro-antd-core-types.umd.js][1] that is mentioned in https://github.com/angular/angular/pull/44245#issuecomment-975912659)).

Therefore, in order to avoid the unnecessary overhead of setting up parent pointers for `sniffModuleFormat()`, I believe it is better to lazily parse the UMD wrapper function if/when it is needed.

[1]: https://unpkg.com/browse/ng-zorro-antd@12.0.2/bundles/ng-zorro-antd-core-types.umd.js

mary-poppins commented:
You can preview d7e1183 at https://pr44245-d7e1183.ngbuilds.io/.

gkalpak commented:
For reference, the `ngcc-validation` tests [passed][1] this time.

[1]: https://app.circleci.com/pipelines/github/angular/ngcc-validation/14445/workflows/d642b590-65a9-44ea-8456-21baaf38eb75

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -521,10 +521,24 @@ export function parseStatementForUmdModule(statement: ts.Statement): UmdModule|n
+    // hand.
+    // For example, if we just want to determine if an entry-point is in CommonJS or UMD format,
+    // trying to parse the wrapper function could potentially throw a (premature) error. By making
+    // the computation of `factoryCalls` lazy, such as error would be thrown later (during an
```
```suggestion
    // the computation of `factoryCalls` lazy, such an error would be thrown later (during an
```

Do we want to have a test for sniffing of UMD bundles that don't follow the expected structure?

gkalpak replied:
LoL - I had added a test, but apparently forgot to push it: https://github.com/angular/angular/pull/44245/commits/b926bc1fa24b221f108088dfd96f215e5031dd2b 😇

mary-poppins commented:
You can preview 0b059f0 at https://pr44245-0b059f0.ngbuilds.io/.

gkalpak committed:
refactor(ngcc): make it easy to support more UMD wrapper function formats

Previously, ngcc could only handle UMD modules whose wrapper function
was implemented as a `ts.ConditionalExpression` (i.e. using a ternary
operator). This is the format emitted by popular bundlers, such as
Rollup. However, this failed to account for a different format, using
`if/else` statements, such as the one [emitted by Webpack][1].

This commit prepares ngcc for supporting different UMD wrapper function
formats by decoupling the operation of parsing the wrapper function body
to capture the various factory function calls and that of operating on
the factory function calls (for example, to read or update their
arguments). In a subsequent commit, this will be used to add support for
the Webpack format.

[1]: https://webpack.js.org/configuration/output/#type-umd

gkalpak committed:
fix(ngcc): support the UMD wrapper function format emitted by Webpack

Previously, ngcc could only handle UMD modules whose wrapper function
was implemented as a `ts.ConditionalExpression` (i.e. using a ternary
operator). This is the format emitted by popular bundlers, such as
Rollup.

This commit adds support for a different format, that uses `if/else`
statements, which is what is [emitted by Webpack][1].

[1]: https://webpack.js.org/configuration/output/#type-umd

Fixes #44019

gkalpak committed:
test(ngcc): use helper function for generating UMD modules for tests

Previously, several ngcc test suites used their own helper to generate
test UMD modules.

This commit switches to using the same helper for generating UMD modules
across test suites. This improves DRYness (ensuring changes/fixes to the
UMD format need only be applied once) and makes it easier to test
different UMD formats in all test suites.

gkalpak committed:
test(ngcc): run more tests against more supported UMD formats

This commit utilizes the infrastructure added in the previous commit to
run more tests against more of the supported UMD formats. This shall
give us more confidence that all aspects of UMD processing work
correctly with the various formats.

gkalpak committed:
fix(ngcc): correctly report error when collecting dependencies of UMD module

Previously, the ngcc `UmdReflectionHost` would throw a misleading error
when trying to collect dependencies of an invalidly formatted UMD
module. This happened because an error would be thrown while trying to
construct the error message for the actual error, by calling `getText()`
on certain TypeScript AST nodes. See
https://github.com/angular/angular/issues/44019#issuecomment-959954121
for a more in-depth explanation.

This commit ensures `getText()` can be safely called on TypeScript AST
nodes when collecting dependencies of UMD modules.

gkalpak committed:
test(ngcc): use `main` property in test `package.json` files for UMD tests

Previously, the mock packages created for `UmdDependencyHost`'s tests,
specified the entry-point as `esm2015`. This does not matter in tests,
since the packages are explicitly passed to the `UmdDependencyHost`
(while in reality the appropriate host would be determined based on the
name of the entry-point property - in this case, detecting the
entry-point as ES2015 and not UMD).

However, in order to avoid confusion, this commit updates the test
packages to use `main` (the default property used for the UMD format in
`package.json` files).

mary-poppins commented:
You can preview 5c69ef7 at https://pr44245-5c69ef7.ngbuilds.io/.

gkalpak committed:
fixup! fix(ngcc): support the UMD wrapper function format emitted by Webpack

mary-poppins commented:
You can preview b2895b2 at https://pr44245-b2895b2.ngbuilds.io/.

dylhunn commented:
This PR was merged into the repository by commit 80776951675e2d55bae5159756ce793888ad986a.
